fs_mark/fs_mark.py

```diff
@@ -0,0 +1,159 @@
+import trunner
+import time
+import re
+import pexpect
+import timing_data as t_data
+import psh.tools.psh as psh
+
+from trunner.ctx import TestContext
+from trunner.dut import Dut
+from trunner.types import TestResult, Status
+
+
+def clean(dut, test_name, ctx: TestContext) -> int:
+    # Dummyfs doesn't need clean since target is rebooted after every test
+    if not ctx.target.rootfs:
+        return 0
+
+    if test_name and ('-k' in test_name or '-L' in test_name or '-F' in test_name):
+        args = test_name.split()
+        args = args[1:]
+
+        def get_dir(x):
+            if args[x[0] - 1] == '-d':
+                return x[1]
+
+        dirs = list(map(lambda x: get_dir(x), enumerate(args)))
+        dirs = [x for x in dirs if x is not None]
+
+        dut.expect(re.escape(ctx.target.shell_prompt), timeout=120)
+        psh._send(dut, f'/bin/fs_mark_clean {" ".join(dirs)}')
+
+        idx = dut.expect([re.escape(ctx.target.shell_prompt),
+                          r".+?Error in remove_dir_recursive.+?",
+                          pexpect.TIMEOUT], timeout=1800)
+
+        if idx == 0:
+            return 0
+        elif idx == 1:
+            return -1
+        elif idx == 2:
+            return -2
+
+    else:
+        return 0
+
+
+def harness(dut: Dut, ctx: TestContext, result: TestResult):
+    target = trunner.ctx.target.name
+    test_status = Status.OK
+
+    test_name = None
+    test_msg = ''
+
+    loop_start = None
+    loop_end = None
+    loop_time = 1200
+    first_loop_done = False
+
+    NAME = r".+?#  (?P<name>(/bin/)?fs_mark.+?)\r+\n"
+    MSG_LINE = r"(?P<line>(\s+\d+){3}.+?)\r+\n"
```

Comment:
Could you clarify whether there have been any issues with regex on any specific target? From my current testing on ia32-generic-qemu, I've managed to capture regex results into groups smoothly, both under no load and under stress. However, I haven't tested this extensively.

My regex:
```Python
    TYPES = [
        "f_use",
        "count",
        "size",
        "files_sec",
        "app_overhead",
        "creatMin",
        "creatAbg",
        "creatMax",
        "writeMin",
        "writeAbg",
        "writeMax",
        "fsyncMin",
        "fsyncAbg",
        "fsyncMax",
        "syncMin",
        "syncAbg",
        "syncMax",
        "closeMin",
        "closeAbg",
        "closeMax",
        "unlinkMin",
        "unlinkAbg",
        "unlinkMax",
    ]
    MSG_LINE = "".join([rf"(?P<{type}>-?\d+\.?\d*)\s+" for type in TYPES]) + r"\r?\n"
```

Code on which that was tested:
```Python
    while True:
        if loop_start and loop_end:
            loop_time = 3 * (loop_end - loop_start)
            loop_start = None
            loop_end = None

        idx = dut.expect([NAME, MSG_LINE, NO_SPC, END, NO_CONT_BLOCK, ERR], timeout=60)
        parsed = dut.match.groupdict()
        print(parsed)

        if idx == 0:
            test_name = parsed["name"]
            loop_start = time.time()
        elif idx == 1:
            pass
```

Output:
```bash
[0J(psh)% /bin/fs_mark -d /fs_mark_test -s 0 -n 10 -v -S 0 -L 10

#  /bin/fs_mark  -d  /fs_mark_test  -s  0  -n  10  -v  -S  0  -L  10 
{'name': '/bin/fs_mark  -d  /fs_mark_test  -s  0  -n  10  -v  -S  0  -L  10 '}
#        Version 3.3, 1 thread(s) starting at Thu Jan 1 00:00:08 1970
#        Sync method: NO SYNC: Test does not issue sync() or fsync() calls.
#        Directories:  no subdirectories used
#        File names: 40 bytes long, (16 initial bytes of time stamp with 24 random bytes at end of name)
#        Files info: size 0 bytes, written with an IO size of 16384 bytes per write
#        App overhead is time in microseconds spent in the test not doing file writing related system calls.
#        All system call times are reported in microseconds.

FSUse%        Count         Size    Files/sec     App Overhead        CREAT (Min/Avg/Max)        WRITE (Min/Avg/Max)        FSYNC (Min/Avg/Max)         SYNC (Min/Avg/Max)        CLOSE (Min/Avg/Max)       UNLINK (Min/Avg/Max)
    27           10            0          1.7           181866   358108   547925   712653     1601     3120     6763        0        0        0        0        0        0     1404     3829    15385        0        0        0
{'f_use': '27', 'count': '10', 'size': '0', 'files_sec': '1.7', 'app_overhead': '181866', 'creatMin': '358108', 'creatAbg': '547925', 'creatMax': '712653', 'writeMin': '1601', 'writeAbg': '3120', 'writeMax': '6763', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1404', 'closeAbg': '3829', 'closeMax': '15385', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
    27           20            0          1.6           237227   419356   590063   842584     1844     3814     7751        0        0        0        0        0        0     1814     4079     9268        0        0        0
    27           30            0          1.4           213842   451503   674516   982013     1746     5525    15782        0        0        0        0        0        0     1719     5564    20754        0        0        0
    27           40            0          1.5           172467   517018   643414   737076     2138     3036     5709        0        0        0        0        0        0     1981     5020    10241        0        0        0
    27           50            0          1.7           156668   501367   577613   700207     2071     3411     6560        0        0        0        0        0        0     1709     2290     3061        0        0        0
    28           60            0          1.5           216290   546834   630984   836860     1656     3969    10902        0        0        0        0        0        0     1937     3399     6637        0        0        0
    28           70            0          1.5           192122   431973   632832   793497     1250     3366    12533        0        0        0        0        0        0     1438     2688     5080        0        0        0
    28           80            0          1.6           139242   479825   622701   775921     1730     4523    10685        0        0        0        0        0        0     1837     3346     9047        0        0        0
{'f_use': '27', 'count': '20', 'size': '0', 'files_sec': '1.6', 'app_overhead': '237227', 'creatMin': '419356', 'creatAbg': '590063', 'creatMax': '842584', 'writeMin': '1844', 'writeAbg': '3814', 'writeMax': '7751', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1814', 'closeAbg': '4079', 'closeMax': '9268', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
{'f_use': '27', 'count': '30', 'size': '0', 'files_sec': '1.4', 'app_overhead': '213842', 'creatMin': '451503', 'creatAbg': '674516', 'creatMax': '982013', 'writeMin': '1746', 'writeAbg': '5525', 'writeMax': '15782', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1719', 'closeAbg': '5564', 'closeMax': '20754', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
{'f_use': '27', 'count': '40', 'size': '0', 'files_sec': '1.5', 'app_overhead': '172467', 'creatMin': '517018', 'creatAbg': '643414', 'creatMax': '737076', 'writeMin': '2138', 'writeAbg': '3036', 'writeMax': '5709', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1981', 'closeAbg': '5020', 'closeMax': '10241', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
{'f_use': '27', 'count': '50', 'size': '0', 'files_sec': '1.7', 'app_overhead': '156668', 'creatMin': '501367', 'creatAbg': '577613', 'creatMax': '700207', 'writeMin': '2071', 'writeAbg': '3411', 'writeMax': '6560', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1709', 'closeAbg': '2290', 'closeMax': '3061', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
{'f_use': '28', 'count': '60', 'size': '0', 'files_sec': '1.5', 'app_overhead': '216290', 'creatMin': '546834', 'creatAbg': '630984', 'creatMax': '836860', 'writeMin': '1656', 'writeAbg': '3969', 'writeMax': '10902', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1937', 'closeAbg': '3399', 'closeMax': '6637', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
{'f_use': '28', 'count': '70', 'size': '0', 'files_sec': '1.5', 'app_overhead': '192122', 'creatMin': '431973', 'creatAbg': '632832', 'creatMax': '793497', 'writeMin': '1250', 'writeAbg': '3366', 'writeMax': '12533', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1438', 'closeAbg': '2688', 'closeMax': '5080', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
{'f_use': '28', 'count': '80', 'size': '0', 'files_sec': '1.6', 'app_overhead': '139242', 'creatMin': '479825', 'creatAbg': '622701', 'creatMax': '775921', 'writeMin': '1730', 'writeAbg': '4523', 'writeMax': '10685', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1837', 'closeAbg': '3346', 'closeMax': '9047', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
    28           90            0          1.5           225379   418652   633652  1053823     1616     3947    18033        0        0        0        0        0        0     2179     4394    10715        0        0        0
{'f_use': '28', 'count': '90', 'size': '0', 'files_sec': '1.5', 'app_overhead': '225379', 'creatMin': '418652', 'creatAbg': '633652', 'creatMax': '1053823', 'writeMin': '1616', 'writeAbg': '3947', 'writeMax': '18033', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '2179', 'closeAbg': '4394', 'closeMax': '10715', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
    28          100            0          1.2           226891   477213   774066  1200110     1676     3605     8894        0        0        0        0        0        0     1573     3729    10540        0        0        0
{'f_use': '28', 'count': '100', 'size': '0', 'files_sec': '1.2', 'app_overhead': '226891', 'creatMin': '477213', 'creatAbg': '774066', 'creatMax': '1200110', 'writeMin': '1676', 'writeAbg': '3605', 'writeMax': '8894', 'fsyncMin': '0', 'fsyncAbg': '0', 'fsyncMax': '0', 'syncMin': '0', 'syncAbg': '0', 'syncMax': '0', 'closeMin': '1573', 'closeAbg': '3729', 'closeMax': '10540', 'unlinkMin': '0', 'unlinkAbg': '0', 'unlinkMax': '0'}
Average Files/sec:          1.0
{}
```

Reply:
The python harness is being run on host - I assume You've been testing the target `ia32-generic-qemu` on your laptop - which is more capable than GA runner (either from GitHub or our RPi).

Regex expressions are hard on CPU (if you don't construct them carefully you could even have O(2^n) complexity!) - so a rule of thumb is to avoid them if they're not necessary (and better use simple ones when not). Please note that intensive regex matching might impact the test total time being measured (not much, but still).

Current approach with splitting the line is better CPU-wise, could probably still be optimized in regards to python syntax, eg:

```python
TYPES = [ "f_use", "count", "size", "files_sec", "app_overhead", "creatMin", "creatAbg", "creatMax", "writeMin", "writeAbg", "writeMax", "fsyncMin", "fsyncAbg", "fsyncMax", "syncMin", "syncAbg", "syncMax", "closeMin", "closeAbg", "closeMax", "unlinkMin", "unlinkAbg", "unlinkMax", ]

# NOTE: used `float` as files_sec is a float number, doesn't really harm other stats
line_dict = dict(zip(TYPES, map(float, line.split())))

# if you need partial-only line_dict - use list splitting, eg:

line_dict = dict(zip(TYPES[20:22], map(float, line.split()[20:22])))

```